cpp/src/interop/from_arrow_host_strings.cu

```diff
@@ -0,0 +1,207 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "arrow_utilities.hpp"
+#include "from_arrow_host.hpp"
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/interop.hpp>
+#include <cudf/detail/utilities/vector_factories.hpp>
+#include <cudf/interop.hpp>
+#include <cudf/strings/detail/strings_children.cuh>
+#include <cudf/strings/detail/strings_column_factories.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/memory_resource.hpp>
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_buffer.hpp>
+#include <rmm/device_uvector.hpp>
+
+#include <thrust/transform.h>
+
+#include <nanoarrow/nanoarrow.h>
+#include <nanoarrow/nanoarrow.hpp>
+
+#include <numeric>
+#include <string>
+#include <vector>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+std::unique_ptr<column> from_arrow_string(ArrowSchemaView* schema,
+                                          ArrowArray const* input,
+                                          std::unique_ptr<rmm::device_buffer>&& mask,
+                                          rmm::cuda_stream_view stream,
+                                          rmm::device_async_resource_ref mr)
+{
+  // offsets column should contain no nulls so we can put nullptr for the bitmask
+  // nulls are tracked in the parent string column itself, not in the offsets
+  void const* offset_buffers[2] = {nullptr, input->buffers[fixed_width_data_buffer_idx]};
+  ArrowArray offsets_array      = {
+         .length     = input->offset + input->length + 1,
+         .null_count = 0,
+         .offset     = 0,
+         .n_buffers  = 2,
+         .n_children = 0,
+         .buffers    = offset_buffers,
+  };
+
+  // chars_column does not contain any nulls, they are tracked by the parent string column
+  // itself instead. So we pass nullptr for the validity bitmask.
+  int64_t const char_data_length = [&]() {
+    if (schema->type == NANOARROW_TYPE_LARGE_STRING) {
+      return reinterpret_cast<int64_t const*>(offset_buffers[1])[input->length + input->offset];
+    } else if (schema->type == NANOARROW_TYPE_STRING) {
+      return static_cast<int64_t>(
+        reinterpret_cast<int32_t const*>(offset_buffers[1])[input->length + input->offset]);
+    } else {
+      CUDF_FAIL("Unsupported string type", cudf::data_type_error);
+    }
+  }();
+  void const* char_buffers[2] = {nullptr, input->buffers[2]};
```

Comment:
The buffer index `fixed_width_data_buffer_idx` above is a named `constexpr` value. Should this `2` also be a named value?

Reply:
I think so. Let me look for that.